### ISSUE TITLE
Create the community standard files

### ISSUE THREAD

**Issue by tukusejssirs** [romcal/romcal#79]
I suggest we should create files that is considered as _community standard_.
- [ ] `contributing.md` (#69, #77);
- [ ] `code_of_conduct.md` (#59);
- [ ] issue template; read https://help.github.com/articles/about-issue-and-pull-request-templates/;
- [ ] pull request template; read https://help.github.com/articles/about-issue-and-pull-request-templates/;
- [ ] `readme.md`, `readme.[lang].md` (#61);
- [ ] `authors.md`; e.g. https://github.com/sinatra/sinatra/blob/master/AUTHORS.md;
- [ ] `contributors`;
- [ ] `history` → `changelog`;
- [ ] (`maintainance`);
- [ ] `version` (#60);
- [ ] `support`; e.g. https://github.com/atom/atom/blob/master/SUPPORT.md;
- [ ] `style_guide`; modify [Angular](https://angular.io/guide/styleguide) to our needs; probably should inporporated in `contributing.md`;
- [ ] `vision.md`; create a list of features to be added in the future (or vision / roadmap); let there be features divided into sub-features (which compose the features);

**Comment by pejulian** [comment]
Hello @tukusejssirs and @emagnier,

In respect to infrastructure enhancements, I've done some work on the deployment process of `romcal` to automate the npm publishing process.

The first step was to create a machine user that can manage automated release activities:

1. Created a new "system" user with the email address (romcalproject@gmail.com)
2. Created a new "machine user" account in GitHub
3. Created a new TravisCI (.org) account with the machine user above.
    - We should only need this travis instance moving forward for running tests, publishing and so on
4. Created a GitHub personal access token to allow TravisCI to push updates to `romcal/romcal`
5. Created an `npm` token to allow TravisCI to publish new packages to `npm`
6. Encrypted the tokens above to be used by travis during deployments
    - These tokens will only be supplied to the build if the travis instance above is used

The second step was to enhance the existing `travis.yml` file:

- Updated `node_js` to support the latest stable release and all LTS releases only
- Run unit tests in the above versions of node 
- Deploy a new version of `romcal` using `semver` with the following constraints
   - The push event is on a branch (not a pull request)
   - The push event is not on a fork of `romcal/romcal`
   - The publish only happens once per build on the latest stable version of node
   - The updated version is pushed back to the branch that the build originated from 

To top it off, I also:

- Streamlined the `npm` hooks in `package.json` where `prepublishOnly` is now the hook that will run the build process before the actual publishing of the package. 
- Added the `files` entry and removed `.npmignore` as the preferred to know what we wish to publish to `npm`
- Added the `semver` package under `devDependencies`

Okay, so let me dive a little into the git flow and semantic versioning process:

We will have 3 fixed branches namely `master`, `test` and `dev` 
Then we will have "feature branches" for bleeding edge features that we want to publish (remember that forks will not be published).

The 3 fixed branches and the feature branch(es) will each be associated to a corresponding npm distribution tag. Distribution tags help us keep the default npm distribution tag (which is `latest`) clean and only populated by stable releases.

The tags are:
- `latest`: The default distribution tag on which publish events from the `master` branch will use
- `beta`: The dist tag for the `test` branch
- `alpha`: The dist tag for the `dev` branch
- `canary`: The dist tag for any feature branch 

Run `npm dist-tags ls romcal` to view all the distribution tags associated to `romcal`.

Next up, semantic versioning:

Each tag will hold published npm versions following a specific semantic versioning syntax:
- `latest`: Holds stable releases, so we will only see `major`, `minor` or `patch` releases here (i.e. 1.3.1, 1.4.0 or 2.0.0)
- `beta`: Holds releases that are close to ready (think of this like our release candidates/rc). Here you will see versions like 1.3.1-beta.1, 2.0.0-beta.0)
- `alpha`: Holds untested but somewhat usable releases. Here you will see versions like 1.3.1-alpha.2, 1.4.0-alpha.13). This could be a nightly build branch if we wanted to.
- `canary`: Holds unstable releases for internal testing purposes. Here you will see versions like 1.3.1-canary.13, 2.0.0-canary.2)

With this, we can now have full flexibility to install the desired version of `romcal` using commands like:

`npm install romcal@alpha` to get the latest alpha release or `npm install romcal@1.3.1-canary.2` to get a specific canary release.

I've created a feature branch named `_F-ci-cd-automation` that holds all these enhancements.

There still remains some todo's like:
- integrating autochangelog into the deployment process
- publishing git tags for master (and perhaps test) branch releases
- finding a way to allow the commiter to override the automatic semantic versioning process and use a user supplied semantic version number (i.e. for major or minor release instances)

**Comment by tukusejssirs** [comment]
Well done, @pejulian! I have nothing else to say with the regard to this work of yours. :smiley:

**Comment by emagnier** [comment]
Awesome! It's a really great addition 👍 
I particularly like the canary tagging on feature branches, as I often want to test my work with different cases studies I'm also working on. This will help us a lot!

**Comment by emagnier** [comment]
Thanks @pejulian for this nice work (# 164), now we can see the publishing process in action, and all the tags on npm are very handy!

There is still an issue in the release process, the `dist` folder contain actually only the` index.js`, while we should have all the javascript files generated by babel (for backend usage), and the `romcal.bundle. min.js` for browsers. Could you take a look on that when you get a chance?

**Comment by tukusejssirs** [comment]
> There is still an issue in the release process, the `dist` folder contain actually only the` index.js`, while we should have all the javascript files generated by babel (for backend usage), and the `romcal.bundle. min.js` for browsers. Could you take a look on that when you get a chance?

Well, I’d vote for having the artifacts in the [GitHub releases](https://github.com/romcal/romcal/releases) only, not in the codebase/repository itself.

**Comment by emagnier** [comment]
@tukusejssirs: Things in the `dist` folder are not in the repository, but automatically generated before publishing a new release, and (normally) integrated in the npm package.

**Comment by tukusejssirs** [comment]
Okay then, sorry I’ve been OT. I mixed up the `dist/` with `bin/`. I’d like to remove the `bin/` (including `bin/romcal`) from the codebase/repo.

**Comment by pejulian** [comment]
> Thanks @pejulian for this nice work (# 164), now we can see the publishing process in action, and all the tags on npm are very handy!
> 
> There is still an issue in the release process, the `dist` folder contain actually only the` index.js`, while we should have all the javascript files generated by babel (for backend usage), and the `romcal.bundle. min.js` for browsers. Could you take a look on that when you get a chance?

Thanks for spotting this @emagnier. I probably should have run `npm pack` locally more often to ensure that the distrubution files matched that of the `/dist` folder. I've fixed the issue and made a build on the `dev` branch now where all the required artifacts should now be in place.

**Comment by romcalproject** [comment]
> Okay then, sorry I’ve been OT. I mixed up the `dist/` with `bin/`. I’d like to remove the `bin/` (including `bin/romcal`) from the codebase/repo.

I've removed the `/bin` folder as I don't see it being used anywhere and can't remember why I added it there in the first place. It was probably for some testing purpose.

**Comment by pejulian** [comment]
Hey all, 

I've gone ahead an removed travis-ci.com as an authorized app on the `romcal` organization.
Moving forward, we can rely on webhooks such as [this](https://github.com/romcal/romcal/settings/hooks) to push CI builds to our travis-ci.org machine user instance. By removing the app, build statuses are now no longer confused and reflecting the _actual_ result of the build.

**Comment by tukusejssirs** [comment]
Nice work, @pejulian!

Just two things:
1. You use two accounts currently, so sometimes you forget switch back to @pejulian  for comments, etc.
2. I (and maybe also @emagnier) don’t have access to romcal repo settings, therefore your link to [webhooks](https://github.com/romcal/romcal/settings/hooks) leads to 404 error. Still I think they are great.

**Comment by tukusejssirs** [comment]
> I've removed the `/bin` folder as I don't see it being used anywhere and can't remember why I added it there in the first place. It was probably for some testing purpose.

Well, It might be used directly (as an external library), but for that I’d rather use a release arfifact (non-compressed).

**Comment by emagnier** [comment]
Also on my side, I don't think `/bin` is used either.

@tukusejssirs npm already manage the release artifacts well, and when you do a `npm install romcal`, you get everything unzipped in the `node_modules`. Moreover, thanks to the @pejulian's work, this package will now contain only the necessary files. This is also the best tooling to switch easily between versions or upgrade to latest.
When releasing a new version, the package should still be zipped since it contains multiple required files. Also I see fewer and fewer libraries or framework offering a direct download. In modern web app/site development, using a package manager (like npm) is generally a widely encouraged practice, so I will prefer to encourage it as well for romcal. 
If there was anything to improve here, it would eventually be to add support for `yarn`, someday. And wrap-up a same generated code for frontend and backend (with lazy load support for  dependencies in the browser), so we get the same entry point to the romcal library from the `package.json`.

**Comment by tukusejssirs** [comment]
@pejulian, @emagnier, do whatever you think is the best. I’ve just thought that it might be a good idea to have the artifacts (like the bundled JS file) in the GitHub releases page, too. If you think it is not that good idea, just don’t do it. :smiley:

> When releasing a new version, the package should still be zipped since it contains multiple required files.

I did not say that the current zip/tar artifact in the GitHub release page should be presented decompressed, but an additional artifact, a non-compressed JS file.

> In modern web app/site development, using a package manager (like npm) is generally a widely encouraged practice, so I will prefer to encourage it as well for romcal.

I didn’t say that the packager should not be the _prefered_ way to manage romcal version(s), but at least for quick tests it might be a good idea to use an officially hosted, non-compress JS file.

---

Just to note: I agree with everything you said/proposed. :smiley: :+1:

**Comment by emagnier** [comment]
Thank you for your reply. Just a clarification: when you say "a non-compressed JS file", actually there isn't one file but several files which depend on each other. Except the one for browsers, but this one should be split as well as I described in my previous comment.

**Comment by tukusejssirs** [comment]
> Just a clarification: when you say "a non-compressed JS file", actually there isn't one file but several files which depend on each other. Except the one for browsers, but this one should be split as well as I described in my previous comment.

Well, I was talking about [dist/romcal.bundle.min.js](https://github.com/romcal/romcal/tree/9546fa064f930f22ab19481557f5d172d40eb452/dist) file, but (from the point of view I described in my previous comment(s)) it might be good idea to make them all available in the uncompressed form.

By the way, in the [GitHub releases](https://github.com/romcal/romcal/releases), the zip/tar compressed files are _not_ artifacts at all; these archives contain the repo only, i.e. therefore there are no artifacts at all.

And by _artifacts_ I mean romcal in JS file(s). :smiley:

**Comment by pejulian** [comment]
> > Just a clarification: when you say "a non-compressed JS file", actually there isn't one file but several files which depend on each other. Except the one for browsers, but this one should be split as well as I described in my previous comment.
> 
> Well, I was talking about [dist/romcal.bundle.min.js](https://github.com/romcal/romcal/tree/9546fa064f930f22ab19481557f5d172d40eb452/dist) file, but (from the point of view I described in my previous comment(s)) it might be good idea to make them all available in the uncompressed form.
> 
> By the way, in the [GitHub releases](https://github.com/romcal/romcal/releases), the zip/tar compressed files are _not_ artifacts at all; these archives contain the repo only, i.e. therefore there are no artifacts at all.
> 
> And by _artifacts_ I mean romcal in JS file(s). 😃

Hello @tukusejssirs,

Here's my take on this subject:

Since we use `npm` to publish our releases, [that](https://www.npmjs.com/package/romcal) should be the go-to place for obtaining (and installing) artefacts. Since the primary objective of `romcal` is to serve as a node module, it is a fair assumption that `npm` should be the reference point for distributions of this library.

GitHub releases are not meant to contain artefacts by nature as they are created from git release tags which are in themselves simply checkpoints in our codebase's commit timeline that we have specified. Therefore, releases should simply imply a "snapshot" as it were, of the codebase at the point the tag associated to it was made.

Having said that, I do understand the concerns you raised. I feel that the right delivery mechanism to achieve this might be a CDN or AWS S3 bucket. A CDN can contain cached versions of static resources that can be linked to in HTML pages. An S3 bucket is simply a container that can be used to hold the uncompressed contents of the dist folder that can be made reachable to the general public. Perhaps in time, I can look into how we can make romcal distributions available by one or both of these channels.

**Comment by pejulian** [comment]
@tukusejssirs @emagnier 

Just a quick update:

If you check out the `_F-community-standards` branch, you'll see auto changelog in action! It updates our `HISTORY.md` automagically on deployments.

Automated tag creation is also available on this branch (albeit, it has been configured to work only on non feature branches). We will get to see it working once the code here is merged to our main branches.

Do let me know your thoughts!

**Comment by tukusejssirs** [comment]
Thank you for your kind comments and your amazing work, @pejulian!

I agree with you and have nothing to add.

As for the auto-changelogging: it is a great feature that eases maintaining romcal! Anyway, do we want to keep the `HISTORY.md` filename? Won’t it be better to rename it to `CHANGELOG.md`? It is not a must-have, just a thought. :smiley:

**Comment by emagnier** [comment]
@pejulian I like the auto changelog! I'm looking forward to see it in other branches!

Regarding the artifacts, a CDN (or S3) for browser usage is a very good idea. My take on it is the stable version is enough. Using beta/alpha versions is still for an advanced usage and not recommended for production, so I might recommand to stick on npm delivery for these advanced versions (and a package manager remains the best tool to play in these non-production-ready versions, especially since the release of these versions are automated and are very easy to install from npm).
For backend usage I'm still deeply against having them outside a package manager like npm.

**Comment by pejulian** [comment]
> Thank you for your kind comments and your amazing work, @pejulian!
> 
> I agree with you and have nothing to add.
> 
> As for the auto-changelogging: it is a great feature that eases maintaining romcal! Anyway, do we want to keep the `HISTORY.md` filename? Won’t it be better to rename it to `CHANGELOG.md`? It is not a must-have, just a thought. 😃

I agree! I have renamed the file to `CHANGELOG.md` to better reflect what the contents of the file represent.

**Comment by pejulian** [comment]
> @pejulian I like the auto changelog! I'm looking forward to see it in other branches!
> 
> Regarding the artifacts, a CDN (or S3) for browser usage is a very good idea. My take on it is the stable version is enough. Using beta/alpha versions is still for an advanced usage and not recommended for production, so I might recommand to stick on npm delivery for these advanced versions (and a package manager remains the best tool to play in these non-production-ready versions, especially since the release of these versions are automated and are very easy to install from npm).
> For backend usage I'm still deeply against having them outside a package manager like npm.

I agree with this as well. We should only deliver static content that is production ready as that will be the easiest way one could gain access to this library (without the need of npm). For the back end, it should be used like any other library for a back end:- installation and management via a package manager (which is NPM in this case).

**Comment by TobiTenno** [comment]
@tukusejssirs @emagnier i don't think vision, maintenance, or support are necessary

- vision is routinely expressed in our discussions and issues
- support is all in here, and attempts to create other communication channels have been shot down already, so no point having a document saying "we only take discussions and issues"
  - if this has changed, let me know and i'll happily make something documenting the current options
  - I know email as discussed at some point, but i'm not sure of the current state
  - it'd be really nice to have a working chat system that we will provide support, discussion, or at least a place to talk. i've proposed discord in the past
- i don't understand the point of the maintenance file, so we should discuss this one

**Comment by TobiTenno** [comment]
closing, given above
